internal/controlplane/handlers_repositories.go

```diff
@@ -99,6 +98,62 @@ func (s *Server) RegisterRepository(
 	}, nil
 }
 
+func (s *Server) repoCreateInfoFromUpstreamRepositoryRef(
+	ctx context.Context,
+	projectID uuid.UUID,
+	providerName string,
+	rep *pb.UpstreamRepositoryRef,
+) (*properties.Properties, *db.Provider, error) {
+	// If the repo owner is missing, GitHub will assume a default value based
+	// on the user's credentials. An explicit check for owner is left out to
+	// avoid breaking backwards compatibility.
+	if rep.GetName() == "" {
+		return nil, nil, util.UserVisibleError(codes.InvalidArgument, "missing repository name")
+	}
+
+	fetchByProps, err := properties.NewProperties(map[string]any{
+		properties.PropertyUpstreamID: fmt.Sprintf("%d", rep.GetRepoId()),
+		properties.PropertyName:       fmt.Sprintf("%s/%s", rep.GetOwner(), rep.GetName()),
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating properties: %w", err)
+	}
+
+	provider, err := s.inferProviderByOwner(ctx, rep.GetOwner(), projectID, providerName)
+	if err != nil {
+		pErr := providers.ErrProviderNotFoundBy{}
+		if errors.As(err, &pErr) {
+			return nil, nil, util.UserVisibleError(codes.NotFound, "no suitable provider found, please enroll a provider")
+		}
+		return nil, nil, status.Errorf(codes.Internal, "cannot get provider: %v", err)
+	}
+
+	return fetchByProps, provider, nil
+}
+
+func (s *Server) repoCreateInfoFromUpstreamEntityRef(
+	ctx context.Context,
+	projectID uuid.UUID,
+	providerName string,
+	entity *pb.UpstreamEntityRef,
+) (*properties.Properties, *db.Provider, error) {
+	inPropsMap := entity.GetProperties().AsMap()
+	fetchByProps, err := properties.NewProperties(inPropsMap)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating properties: %w", err)
+	}
+
+	provider, err := s.providerStore.GetByName(ctx, projectID, providerName)
```

Comment:
Shouldn't this search in the hierarchy? I got confused by  repoCreateInfoFromUpstreamRepositoryRef calling inferProvider and this call going straight for GetByName in the project.

Reply:
I think the provider store already uses the hierarchy, IIRC